src/test/change_stream.rs

```diff
@@ -0,0 +1,656 @@
+use bson::{doc, Bson, Document};
+use futures_util::StreamExt;
+use semver::VersionReq;
+
+use crate::{
+    change_stream::{
+        event::{ChangeStreamEvent, OperationType},
+        options::ChangeStreamOptions,
+        ChangeStream,
+    },
+    event::command::{CommandStartedEvent, CommandSucceededEvent},
+    test::{CommandEvent, FailCommandOptions, FailPoint, FailPointMode},
+    Collection,
+};
+
+use super::{EventClient, TestClient, CLIENT_OPTIONS, LOCK};
+
+type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;
+
+async fn init_stream(
+    coll_name: &str,
+    direct_connection: bool,
+) -> Result<
+    Option<(
+        EventClient,
+        Collection<Document>,
+        ChangeStream<ChangeStreamEvent<Document>>,
+    )>,
+> {
+    let init_client = TestClient::new().await;
+    if !init_client.is_replica_set() && !init_client.is_sharded() {
+        println!("skipping change stream test on unsupported topology");
+        return Ok(None);
+    }
+    if !init_client.supports_fail_command() {
+        println!("skipping change stream test on version without fail commands");
+        return Ok(None);
+    }
+
+    let mut options = CLIENT_OPTIONS.clone();
+    // Direct connection is needed for reliable behavior with fail points.
+    if direct_connection && init_client.is_sharded() {
+        options.direct_connection = Some(true);
+        options.hosts.drain(1..);
+    }
+    let client = EventClient::with_options(options).await;
+    let db = client.database("change_stream_tests");
+    let coll = db.collection::<Document>(coll_name);
+    coll.drop(None).await?;
+    let stream = coll.watch(None, None).await?;
+    Ok(Some((client, coll, stream)))
+}
+
+/// Prose test 1: ChangeStream must continuously track the last seen resumeToken
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn tracks_resume_token() -> Result<()> {
+    let _guard = LOCK.run_concurrently().await;
+
+    let (client, coll, mut stream) = match init_stream("track_resume_token", false).await? {
+        Some(t) => t,
+        None => return Ok(()),
```

Comment:
can we log that we're skipping due to topology here? (in the spirit of #523)

Reply:
`init_stream` logs when it returns None :)